### PR TITLE
add webapp ssh

### PR DESCRIPTION
**Pull request by patricklee2**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
-
+# pylint: disable=C0302
```
Comment (williexu) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py]: Please use the name of the message for more clarity

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py [hunk 2]
```diff
@@ -331,6 +331,10 @@ def load_arguments(self, _):
         c.argument('sku', arg_type=sku_arg_type)
         c.argument('dryrun', help="show summary of the create and deploy operation instead of executing it", default=False, action='store_true')
 
+    with self.argument_context('webapp ssh') as c:
+        c.argument('slot', options_list=['--slot', '-s'], help='slot to restore the deleted content to')
+        c.argument('container', options_list=['--container', '-c'], help='the DNS name of container in multi-container web app (currently not supported)')
```
Comment (williexu) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py]: What exactly is not currently supported? It almost sounds like the parameter is not supported.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 3]
```diff
@@ -2192,3 +2193,81 @@ def _ping_scm_site(cmd, resource_group, name):
     import urllib3
     authorization = urllib3.util.make_headers(basic_auth='{}:{}'.format(user_name, password))
     requests.get(scm_url + '/api/settings', headers=authorization)
+
+
+def _check_for_ready_tunnel(remote_debugging, tunnel_server):
+    default_port = tunnel_server.is_port_set_to_default()
+    if default_port is not remote_debugging:
+        return True
+    return False
```
Comment (williexu) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: `return default_port is not remote_debugging`

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 4]
```diff
@@ -2192,3 +2193,81 @@ def _ping_scm_site(cmd, resource_group, name):
...
+    is_linux = webapp.reserved
+    if not is_linux:
+        logger.error("Only Linux App Service Plans supported, Found a Windows App Service Plan")
+        return
+    import time
+    profiles = list_publish_profiles(cmd, resource_group_name, name, slot)
+    user_name = next(p['userName'] for p in profiles)
+    user_password = next(p['userPWD'] for p in profiles)
+
+    from .tunnel import TunnelServer
```
Comment (williexu) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: These imports can probably go at the start of the file unless they really affect perf.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tunnel.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                else:
+                    logger.warning('Client disconnected %s', index)
+                    client.close()
+                    ws_socket.close()
+                    break
+            except:
+                traceback.print_exc(file=sys.stdout)
+                client.close()
+                ws_socket.close()
+                return False
```
Comment (williexu) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tunnel.py]: Is this boolean value used? Currently, this function only ever returns False or None

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tunnel.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                else:
+                    logger.info('Client disconnected!, index: %s', index)
+                    client.close()
+                    ws_socket.close()
+                    break
+            except:
+                traceback.print_exc(file=sys.stdout)
+                client.close()
+                ws_socket.close()
+                return False
```
Comment (williexu) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tunnel.py]: Is this boolean return value used? It's best to have consistent return values, either a None, or a Bool

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py (hunk 2 above)
Comment (patricklee2) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py]: yes, its not supported yet, ill remove it

**Comment by patricklee2** [comment]
@williexu made updates, please review again.

**Comment by patricklee2** [comment]
@yugangw-msft  @panchagnula when you get some time, please review this

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py [hunk 7]
```diff
@@ -480,6 +480,21 @@ def test_linux_webapp(self, resource_group):
...
+        runtime = 'node|8.11'
+        plan = self.create_random_name(prefix='webapp-ssh-plan', length=24)
+        webapp = self.create_random_name(prefix='webapp-ssh', length=24)
+        self.cmd('appservice plan create -g {} -n {} --sku S1 --is-linux' .format(resource_group, plan))
+        self.cmd('webapp create -g {} -n {} --plan {} --runtime {}'.format(resource_group, webapp, plan, runtime))
+        time.sleep(30)
+        requests.get('http://{}.azurewebsites.net'.format(webapp), timeout=240)
+        time.sleep(30)
+        self.cmd('webapp ssh -g {} -n {}'.format(resource_group, webapp))
+        time.sleep(30)
```
Comment (patricklee2) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py]: @panchagnula added test, but its really hard to test this

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 8]
```diff
@@ -2231,3 +2233,72 @@ def _ping_scm_site(cmd, resource_group, name):
...
+
+
+def create_tunnel(cmd, resource_group_name, name, port=None, slot=None):
+    webapp = show_webapp(cmd, resource_group_name, name, slot)
+    is_linux = webapp.reserved
+    if not is_linux:
+        logger.error("Only Linux App Service Plans supported, Found a Windows App Service Plan")
+        return
+    profiles = list_publish_profiles(cmd, resource_group_name, name, slot)
+    user_name = next(p['userName'] for p in profiles)
```
[the remark this answers is not in the document]
Comment (patricklee2) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: this was a mistake, these creds are not used

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 9]
```diff
@@ -2231,3 +2233,72 @@ def _ping_scm_site(cmd, resource_group, name):
...
+            time.sleep(1)
+            if print_warnings:
+                logger.warning('.')
+            if _check_for_ready_tunnel(tunnel_server):
+                break
+
+
+def _start_tunnel(tunnel_server):
+    _wait_for_tunnel(tunnel_server, True)
+    logger.warning('SSH is available { username: root, password: Docker! }')
```
[the remark this answers is not in the document]
Comment (patricklee2) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: we have no plans to change these login creds

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 10]
```diff
@@ -2192,3 +2193,69 @@ def _ping_scm_site(cmd, resource_group, name):
...
+
+def _check_for_ready_tunnel(tunnel_server):
+    return tunnel_server.is_port_set_to_default()
+
+
+def create_tunnel(cmd, resource_group_name, name, port=None, slot=None):
+    webapp = show_webapp(cmd, resource_group_name, name, slot)
+    is_linux = webapp.reserved
+    if not is_linux:
+        logger.error("Only Linux App Service Plans supported, Found a Windows App Service Plan")
```
[the remark this answers is not in the document]
Comment (patricklee2) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: added a throw

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py [hunk 11]
```diff
@@ -347,6 +347,9 @@ def load_arguments(self, _):
         c.argument('sku', arg_type=sku_arg_type)
         c.argument('dryrun', help="show summary of the create and deploy operation instead of executing it", default=False, action='store_true')
 
+    with self.argument_context('webapp ssh') as c:
+        c.argument('slot', options_list=['--slot', '-s'], help='slot to restore the deleted content to')
```
[the remark this answers is not in the document]
Comment (patricklee2) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py]: fixed

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py [hunk 12]
```diff
@@ -659,6 +657,18 @@
             az webapp create -g MyResourceGroup -p MyPlan -n MyUniqueAppName --runtime "node|6.2" --deployment-local-git
 """
 
+helps['webapp ssh'] = """
+    type: command
+    short-summary: (Preview) SSH command establishes a ssh session to the web container and developer would get a shell terminal remotely.
+                   Developer should be able to use the common SSH clients such as the native SSH client from commandline on Linux or Mac,
+                   and PuTTY on Windows.  For blessed images, an openssh service is installed and started on the web container by default.
+                   For WAfC, we requires developer to include and start the openssh service in custom container(s).
```
[the remark this answers is not in the document]
Comment (patricklee2) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py]: a bad copy-paste, fixed

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 13]
```diff
@@ -2231,3 +2233,72 @@ def _ping_scm_site(cmd, resource_group, name):
...
+
+
+def _start_ssh(host_name, port):
+    subprocess.call("ssh -o StrictHostKeyChecking=no root@{} -p {}".format(host_name, port), shell=True)
+
+
+def ssh_webapp(cmd, resource_group_name, name, slot=None):  # pylint: disable=too-many-statements
+    import platform
+    if platform.system() == "Windows":
+        logger.warning('webapp ssh is only supported on linux and mac')
```
[the remark this answers is not in the document]
Comment (patricklee2) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: added a throw

**Pull request merged**